### ISSUE TITLE
Compute type parameters for calls

### ISSUE THREAD

**Issue by lars-reimann** [Safe-DS/DSL#861]
For calls, we must infer type parameters of classes (constructor call) and functions using the passed parameters. We must also ensure and test that the inference of the type of lambdas works properly when type parameters are involved.

Examples from Kotlin:

```kt
fun <T>f1(a: T, b: T): T {
    return a
}

fun <T>f2(a: () -> T, b: () -> T): T {
    return a()
}

fun <T>f3(a: (p: T) -> T, b: T): T {
    return a(b)
}

fun main() {
    // lowest common supertype
    val a1 = f1(1, 2) // Int
    val a2 = f1(1, 2.0) // {Comparable<*> & Number}
    val a3 = f1(1, "") // {Comparable<*> & java.io.Serializable}

    // lowest common supertype
    val b1 = f2({ 1 }, { 2 }) // Int
    val b2 = f2({ 1 }, { 2.0 }) // {Comparable<*> & Number}
    val b3 = f2({ 1 }, { "" }) // {Comparable<*> & java.io.Serializable}

    // lowest common supertype
    val c1 = f3({ it }, 3) // Int
    val c2 = f3({ it }, 2.0) // Double
    val c3 = f3({ it }, "") // String

    val d1 = f3({ p -> "" }, 3) // Int - with errors (would actually work if we replace `T` with Any)
    val d2 = f3({ p -> "" }, 2.0) // Double - with errors (would actually work if we replace `T` with Any)
    val d3 = f3({ p -> "" }, "") // String
    
    val e1 = f3({ p: Int -> p }, 3) // Int
    val e2 = f3({ p: Int -> p }, 2.0) // Int - with errors
    val e3 = f3({ p: Int -> p }, "") // Int - with errors

    val f1 = f3({ p: Int -> "" }, 3) // {Comparable<*> & java.io.Serializable} - with errors
    val f2 = f3({ p: Int -> "" }, 2.0) // {Comparable<*> & java.io.Serializable} - with errors
    val f3 = f3({ p: Int -> "" }, "") // String - with errors
}
```

Potential implementation (assuming a single parameter):

Given a parameter type $T_1$ that contains an unsubstituted type parameter $V$ and the corresponding argument type $T_2$:

* Traverse $T_1$ and $T_2$ in parallel from left to right. Skip any containers that don't match.
* Keep track of whether we are in a covariant context (top-level of constructor, callable type results at top level, inside covariant containers at top level, ...), contravariant context (callable type parameters at top level, inside contravariant containers at top level, ...), or invariant context (invariant container). This computation is already done to check usages of type parameters. Maybe we can reuse it for both tasks. Henceforth, we refer to this as the "context variance".
* Keep track of an aggregated type $Agg$ for $V$. Initially, this is `undefined`.
* Keep track of the "remaining variance" of $Agg$. Initially, this is bivariant.
* If we find an occurrence of $V$ in $T_1$, let the corresponding type in $T_2$ be $Cur$. We must be cautious about lambdas here, which might not specify types for their parameter.
    * If $Cur$ is equal to $V$, continue. This can happen for lambda parameters without manifest types. 
    * If $Agg$ is `undefined`: Set $Agg$ to $Cur$. Set the "remaining variance" to the "context variance".
    * If $Agg$ is defined:
        * "remaining variance" is invariant: Cur must equal $Agg$. If it's not, set $V$ to $Agg$ and stop.
        * "remaining variance" is contravariant:
            * "context variance" is contravariant: Set $Agg$ to the highest common subtype of $Agg$ and $Cur$.
            * "context variance" is covariant/invariant: $Cur$ must equal $Agg$. If it's not, set $V$ to $Agg$ and stop. Otherwise, set the "remaining variance" to invariant.
        * "remaining variance" is covariant:
            * "context variance" is covariant: Set $Agg$ to the lowest common supertype of $Agg$ and $Cur$.
            * "context variance" is contravariant/invariant: $Cur$ must equal $Agg$. If it's not, set $V$ to $Agg$ and stop. Otherwise, set the "remaining variance" to invariant.
* If $Agg$ is `undefined`, set $V$ to `Nothing`. Otherwise, set $V$ to $Agg$.

➡️ Improvement: We can stop once the remaining variance is invariant. $Agg$ can never change afterward. We can also omit the equality check when "remaining variance" and "context variance" don't match.

➡️ If the inferred type for $V$ is not a subtype of its upper bound, we could also replace it with the upper bound. Then we can show errors for #915 on the arguments that fail the bounds check without extra work.
